File: LogicServices/LinearRegression.cs
using System.Linq;
using System.Text.RegularExpressions;

namespace LogicServices
{
   public class XYPair
   {
      public double x { get; set; }
      public double y { get; set; }
   }

   public class LinearRegression
   {
      // ---------------------------------------------------------------------
      // Extraction/Parsing Functions
      // ---------------------------------------------------------------------
      private List<double> ParseForDoublesLine(string input)
      {
         List<double> values = new List<double>();

         // First split the input by commas.
         string[] ary = Regex.Split(input, ",");

         foreach (var s in ary)
         {
            if (s != "")
               // TODO: The standard parser for double strips commas. So 3,3,0 becomes 330. Find a fix for this.
               values.Add(double.Parse(s));
         }

         return values;
      }

      public bool IsDataValidForLinearRegression (string input, out List<XYPair> values)
      {
         values = new List<XYPair> ();

         try
         {
            // First split the input at newlines.
            string[] ary = Regex.Split(input, "\r\n|\r|\n");

            foreach (var s in ary)
            {
               if (s != "")
               {
                  // Expect two values per line (comma separated).
                  string[] xy = s.Split(',');

                  if (xy.Length != 2)
                     return (false);

                  XYPair pair = new XYPair();
                  pair.x = double.Parse(xy[0]);
                  pair.y = double.Parse(xy[1]);
                  values.Add(pair);
               }
            }
         }
         catch
         {
            return (false);
         }

         if (values.Count == 0)
            return false;

         return true;
      }

      public bool IsDataValidForYValue (string input, out double x, out double m, out double b)
      {
         x = 0;
         m = 0;
         b = 0;
         try
         {
            List<double> values = ParseForDoublesLine(input);

            if ((values == null) || (values.Count != 3))
               return false;
            else
            {
               x = values[0];
               m = values[1];
               b = values[2];
            }
            return (true);
         }
         catch
         {
            return (false);
         }
      }

      public string CalculateLinearRegression(List<XYPair> values)
      {
         // Find averages for x and y
         double sumx = 0, sumy = 0;
         foreach (XYPair pair in values)
         {
            sumx += pair.x;
            sumy += pair.y;
         }
         double avgx = sumx / values.Count;
         double avgy = sumy / values.Count;

         // Find sum of the square of differences
         double sumsqx = 0;
         double mtemp = 0;
         foreach (XYPair pair in values)
         {
            sumsqx += Math.Pow((avgx - pair.x), 2);
            double t = (avgx - pair.x) * (avgy - pair.y);
            mtemp += t;
         }
         double m = mtemp / sumsqx;
         double b = avgy - (m * avgx);

         return ("y = " + m + "x + " + b);
      }

      public double CalculateYValue(double x, double m, double b)
      {
         double y = m * x + b;
         return (y);
      }
   }
}

File: calculator/src/main/java/com/calculatorlogic/LinearRegression.java
package com.calculatorlogic;

import ch.qos.logback.core.net.SyslogOutputStream;
import com.calculator.models.XMBInput;

import java.util.ArrayList;
import java.util.List;

public class LinearRegression
{
   // ---------------------------------------------------------------------
   // Extraction/Parsing Functions
   // ---------------------------------------------------------------------
   private List<Double> ParseForDoublesLine(String input)
   {
      List<Double> values = new ArrayList<Double>();

      // First split the input by commas.
      String blah = input.replace("\"", "");
      String[] ary = blah.split(",");

      for (String s : ary)
      {
         Double d = Double.valueOf(s);
         values.add(d);
      }

      return values;
   }

   public List<XYPair> IsDataValidForLinearRegression (String input)
   {
      List<XYPair> values = new ArrayList<XYPair> ();

      try
      {
         // First split the input at newlines.
         String blah = input.replace("\"", "");
         String blah2 = blah.replace("\\", "");
         String blah3 = blah2.replace("n", "%");
         String[] ary = blah3.split("%");

         for (String s: ary)
         {
            if (!s.isEmpty())
            {
               // Expect two values per line (comma separated).
               String[] xy = s.split(",");

               if (xy.length != 2)
               {
                  //System.out.println("lr length not 2");
                  return (null);
               }

               XYPair pair = new XYPair();
               pair.x = Double.valueOf(xy[0]);
               pair.y = Double.valueOf(xy[1]);
               values.add(pair);
            }
         }
      }
      catch(Exception e)
      {
         //System.out.println("lr exception");
         return (null);
      }

      if (values.isEmpty())
      {
         //System.out.println("lr size is 0");
         return null;
      }
      //System.out.println("lr good");
      return values;
   }


   public XMBInput IsDataValidForYValue (String input)
   {
      XMBInput d = new XMBInput ();
      try
      {
         List<Double> values = ParseForDoublesLine(input);

         //System.out.println(values);
         //if (values == null)
         //{
         //   //System.out.println("y value bad");
         //   return (null);
         //}
         if (values.size() != 3)
         {
            return (null);
         }
         else
         {
            d.x = values.get(0);
            d.m = values.get(1);
            d.b = values.get(2);
            //System.out.println("y value good");
         }
      }

      catch (Exception e)
      {
         //System.out.println("y value exception");
         return (null);
      }
      return (d);
   }

   public String CalculateLinearRegression(List<XYPair> values)
   {
      if ((values == null) || (values.isEmpty()))
         return null;
      // Find averages for x and y
      double sumx = 0.0, sumy = 0.0;
      for (XYPair pair : values)
      {
         sumx += pair.x;
         sumy += pair.y;
      }
      double avgx = sumx / values.size();
      double avgy = sumy / values.size();

      // Find sum of the square of differences
      double sumsqx = 0.0;
      double mtemp = 0.0;
      for (XYPair pair : values)
      {
         sumsqx += Math.pow((avgx - pair.x), 2);
         double t = (avgx - pair.x) * (avgy - pair.y);
         //System.out.println(t);
         mtemp += t;
      }
      //System.out.println(sumsqx);
      double m = mtemp / sumsqx;
      double b = avgy - (m * avgx);

      return ("y = " + m + "x + " + b);
   }

   public Double CalculateYValue(XMBInput input)
   {
      if (input == null)
           return null;
      double y = input.m * input.x + input.b;
      return (y);
   }
}
